src/rdkafka_offset.c

```diff
@@ -808,37 +859,353 @@ void rd_kafka_offset_reset(rd_kafka_toppar_t *rktp,
+                        /* Reset the epoch to -1 since it can't be used with
+                         * older brokers. */
+                        rktp->rktp_next_fetch_start.leader_epoch = -1;
+                        rd_kafka_toppar_set_fetch_state(
+                            rktp, RD_KAFKA_TOPPAR_FETCH_ACTIVE);
+                        goto done;
+
+                } else if (err == RD_KAFKA_RESP_ERR__OUTDATED) {
+                        /* Partition state has changed, this response
+                         * is outdated. */
+                        goto done;
+                }
+
+                actions = rd_kafka_err_action(
+                    rkb, err, request, RD_KAFKA_ERR_ACTION_REFRESH,
+                    RD_KAFKA_RESP_ERR_UNKNOWN_LEADER_EPOCH,
+                    RD_KAFKA_ERR_ACTION_REFRESH,
+                    RD_KAFKA_RESP_ERR_FENCED_LEADER_EPOCH,
+                    RD_KAFKA_ERR_ACTION_REFRESH,
+                    RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART,
+                    RD_KAFKA_ERR_ACTION_REFRESH,
+                    RD_KAFKA_RESP_ERR_OFFSET_NOT_AVAILABLE,
+                    RD_KAFKA_ERR_ACTION_REFRESH,
+                    RD_KAFKA_RESP_ERR_KAFKA_STORAGE_ERROR,
+                    RD_KAFKA_ERR_ACTION_END);
+
+
+                if (actions & RD_KAFKA_ERR_ACTION_REFRESH)
+                        /* Metadata refresh is ongoing, so force it */
+                        rd_kafka_topic_leader_query0(rk, rktp->rktp_rkt, 1,
+                                                     rd_true /* force */);
+
+                if (actions & RD_KAFKA_ERR_ACTION_RETRY) {
+                        /* No need for refcnt on rktp for timer opaque
+                         * since the timer resides on the rktp and will be
+                         * stopped on toppar remove. */
+                        rd_kafka_timer_start_oneshot(
+                            &rk->rk_timers, &rktp->rktp_validate_tmr, rd_false,
+                            500 * 1000 /* 500ms */,
+                            rd_kafka_offset_validate_tmr_cb, rktp);
+                        goto done;
+                }
+
+                if (!(actions & RD_KAFKA_ERR_ACTION_REFRESH)) {
+                        /* Permanent error */
+                        rd_kafka_offset_reset(
+                            rktp, rd_kafka_broker_id(rkb),
+                            RD_KAFKA_FETCH_POS(RD_KAFKA_OFFSET_INVALID,
+                                               rktp->rktp_leader_epoch),
+                            RD_KAFKA_RESP_ERR__LOG_TRUNCATION,
+                            "Unable to validate offset and epoch: %s",
+                            rd_kafka_err2str(err));
+                }
+                goto done;
+        }
+
+
+        rktpar     = &parts->elems[0];
+        end_offset = rktpar->offset;
+        end_offset_leader_epoch =
+            rd_kafka_topic_partition_get_leader_epoch(rktpar);
+
+        if (end_offset < 0 || end_offset_leader_epoch < 0) {
```

Comment:
Question: just to confirm, this case behaves similar to how it is currently, right? So for example, if auto.offset.reset = earliest and this if condition is true, we will reset the offset to as far back as possible?

Reply:
This should happen only if all the previous and current epochs have been removed by retention policy. In this case it's ok to call a reset following the configured policy. I put a more descriptive message.